### DISCUSSION TITLE
Better integration with typechain

### DISCUSSION THREAD

**Discussion by jrocca82** [wevm/viem#1292] [Idea / Feature Request] [answered]
I've noticed this package doesn't always integrate well with typechain and the use of contract factories. For example, instead of importing an abi JSON file, whiich can be annoying, it is common for developers to use Contract__factory.createInterface(). This ensures that any updates to the contracts also updates the abi on compiling without having to reimport an abi json.

When I try to do this with viem, however, the types of viem's ABI and typechain's createInterface() function create issues in typescript.

 It would be great if viem could improve its integration with typechain/typescript to fix issues like this and allow for the use of contract factories from typechain.

**Comment by jxom** [comment, marked as answer]
viem uses [ABIType](https://abitype.dev/) for type inference, and comes out-of-the-box, meaning you do not need TypeChain. Check out the [Type Inference](https://viem.sh/docs/typescript.html#type-inference) section on the docs for more info.

Also related: https://github.com/dethcrypto/TypeChain/issues/844#issuecomment-1574991501
